src/cluster.py

```diff
@@ -0,0 +1,212 @@
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""The cluster module manages cluster-wide operations."""
+
+import json
+import logging
+import subprocess
+import uuid
+from typing import Tuple
+
+import ops.charm
+import ops_sunbeam.guard as sunbeam_guard
+import tenacity
+from charms.operator_libs_linux.v2 import snap
+
+import charm
+import microceph
+import relation_handlers
+from microceph import CephHealth, CephStatus
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterNodes(ops.framework.Object):
+    """ClusterNodes manages adding and joining nodes to the microceph cluster."""
+
+    def __init__(self, charm: "charm.MicroCephCharm"):
+        super().__init__(charm, "cluster-nodes")
+        self.charm = charm
+
+    def add_node_to_cluster(self, event: ops.framework.EventBase) -> None:
+        """Add node to microceph cluster."""
+        if not event.unit:
+            return
+        cmd = ["sudo", "microceph", "cluster", "add", event.unit.name]
+        try:
+            logger.debug(f'Running command {" ".join(cmd)}')
+            process = subprocess.run(cmd, capture_output=True, text=True, check=True, timeout=180)
+            logger.debug(f"Command finished. stdout={process.stdout}, " f"stderr={process.stderr}")
```

Comment:
I think parts of this could be moved to `microceph.py` (it already has `remove_cluster_member`) and this method could function as the try/except wrapper around it. The `run_cmd` method in `microceph.py` already does debug logging for exec and error logging for failures.

Reply:
The exception handling for this around timeouts is a bit different, so that won't work

Reply:
umm, the `run_cmd` method is supposed to be transparent (with builtin logging) since it re-raises the exception (shown below), I think that would work with the outer try/except in this method. :
```
try:
        process = subprocess.run(cmd, capture_output=True, text=True, check=True, timeout=180)
        logger.debug(f"Command {' '.join(cmd)} finished; Output: {process.stdout}")
        return process.stdout
except subprocess.CalledProcessError as e:
        logger.error(f"Failed executing cmd: {cmd}, error: {e.stderr}")
        raise e
```